File: src/pending-event.ts
import type { ImmutableClient } from './client.js';
import type { Actor, EventPayload, EventResponse, Target } from './types.js';

export class PendingEvent {
  private readonly actorData: Actor;
  private idempotencyKeyValue?: string;
  private versionValue?: number;
  private sessionValue?: string;
  private targetsValue: Target[] = [];
  private actionCategoryValue?: string;

  constructor(
    private readonly client: ImmutableClient,
    actor: Actor,
  ) {
    this.actorData = actor;
  }

  idempotencyKey(key: string): this {
    this.idempotencyKeyValue = key;
    return this;
  }

  version(v: number): this {
    this.versionValue = v;
    return this;
  }

  session(id: string): this {
    this.sessionValue = id;
    return this;
  }

  actionCategory(category: string): this {
    this.actionCategoryValue = category;
    return this;
  }

  target(type: string, id?: string, name?: string, metadata?: Record<string, unknown>): this {
    this.targetsValue.push({ type, id, name, metadata });
    return this;
  }

  targets(targets: Target[]): this {
    this.targetsValue = targets;
    return this;
  }

  async track(
    action: string,
    resource?: string | { type: string; id: string; name?: string },
    metadata?: Record<string, unknown>,
  ): Promise<EventResponse> {
    const payload: EventPayload = {
      actor_id: this.actorData.id,
      action,
    };

    if (this.actorData.name) payload.actor_name = this.actorData.name;
    if (this.actorData.type) payload.actor_type = this.actorData.type;

    if (typeof resource === 'string') {
      payload.resource = resource;
    } else if (resource) {
      payload.resource = resource.type;
      payload.resource_id = resource.id;
      if (resource.name) payload.resource_name = resource.name;
    }

    if (this.actionCategoryValue) payload.action_category = this.actionCategoryValue;
    if (metadata && Object.keys(metadata).length > 0) payload.metadata = metadata;
    if (this.targetsValue.length > 0) payload.targets = this.targetsValue;
    if (this.idempotencyKeyValue) payload.idempotency_key = this.idempotencyKeyValue;
    if (this.versionValue) payload.version = this.versionValue;
    if (this.sessionValue) payload.session_id = this.sessionValue;

    return this.client.track(payload);
  }
}

File: src/getimmutable/pending_event.py
from __future__ import annotations

from typing import TYPE_CHECKING, Any

if TYPE_CHECKING:
    from .client import ImmutableClient
    from .types import EventResponse


class PendingEvent:
    """Fluent builder for constructing and sending an audit event."""

    def __init__(self, client: ImmutableClient, actor: dict[str, Any]) -> None:
        self._client = client
        self._actor = actor
        self._idempotency_key: str | None = None
        self._version: int | None = None
        self._session_id: str | None = None
        self._action_category: str | None = None
        self._targets: list[dict[str, Any]] = []

    def idempotency_key(self, key: str) -> PendingEvent:
        """Set an idempotency key to prevent duplicate events."""
        self._idempotency_key = key
        return self

    def version(self, v: int) -> PendingEvent:
        """Set the event schema version."""
        self._version = v
        return self

    def session(self, session_id: str) -> PendingEvent:
        """Set the session identifier."""
        self._session_id = session_id
        return self

    def action_category(self, category: str) -> PendingEvent:
        """Set the action category."""
        self._action_category = category
        return self

    def target(
        self,
        type: str,
        id: str | None = None,
        name: str | None = None,
        metadata: dict[str, Any] | None = None,
    ) -> PendingEvent:
        """Add a target resource to this event."""
        t: dict[str, Any] = {"type": type}
        if id is not None:
            t["id"] = id
        if name is not None:
            t["name"] = name
        if metadata is not None:
            t["metadata"] = metadata
        self._targets.append(t)
        return self

    def targets(self, targets: list[dict[str, Any]]) -> PendingEvent:
        """Set all targets at once."""
        self._targets = targets
        return self

    def track(
        self,
        action: str,
        resource: str | dict[str, Any] | None = None,
        metadata: dict[str, Any] | None = None,
    ) -> EventResponse:
        """Send the event to the Immutable API.

        Args:
            action: The action name (e.g. "document.created").
            resource: A resource type string, or a dict with type, id, and optionally name.
            metadata: Additional key-value metadata.
        """
        payload: dict[str, Any] = {
            "actor_id": str(self._actor["id"]),
            "action": action,
        }

        if self._actor.get("name"):
            payload["actor_name"] = self._actor["name"]
        if self._actor.get("type"):
            payload["actor_type"] = self._actor["type"]
        if isinstance(resource, dict):
            payload["resource"] = resource.get("type", "")
            if resource.get("id"):
                payload["resource_id"] = resource["id"]
            if resource.get("name"):
                payload["resource_name"] = resource["name"]
        elif resource is not None:
            payload["resource"] = resource
        if metadata and len(metadata) > 0:
            payload["metadata"] = metadata
        if self._action_category:
            payload["action_category"] = self._action_category
        if self._idempotency_key:
            payload["idempotency_key"] = self._idempotency_key
        if self._version is not None:
            payload["version"] = self._version
        if self._session_id:
            payload["session_id"] = self._session_id
        if self._targets:
            payload["targets"] = self._targets

        return self._client.track(payload)
